kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -282,6 +379,277 @@ class KotlinJsonAdapterFactory : JsonAdapter.Factory {
+  require(!kotlinMetadataName.isLocal) {
+    "Local/anonymous classes are not supported!"
+  }
+  // Top-level: package/of/class/MyClass
+  // Nested A:  package/of/class/MyClass.NestedClass
+  val simpleName = kotlinMetadataName.substringAfterLast(
+      '/', // Drop the package name, e.g. "package/of/class/"
+      '.' // Drop any enclosing classes, e.g. "MyClass."
+  )
+  val packageName = kotlinMetadataName.substringBeforeLast("/", missingDelimiterValue = "")
+  val simpleNames = kotlinMetadataName.removeSuffix(simpleName)
+      .removeSuffix(".") // Trailing "." if any
+      .removePrefix(packageName)
+      .removePrefix("/")
+      .let {
+        if (it.isNotEmpty()) {
+          it.split(".")
+        } else {
+          // Don't split, otherwise we end up with an empty string as the first element!
+          emptyList()
+        }
+      }
+      .plus(simpleName)
+
+  return "${packageName.replace("/", ".")}.${simpleNames.joinToString(".")}"
+}
+
+@Suppress("SameParameterValue")
+private fun String.substringAfterLast(vararg delimiters: Char): String {
+  val index = lastIndexOfAny(delimiters)
+  return if (index == -1) this else substring(index + 1, length)
+}
+
+private fun defaultPrimitiveValue(type: Type): Any? =
+    if (type is Class<*> && type.isPrimitive) {
+      when (type) {
+        Boolean::class.java -> false
+        Char::class.java -> 0.toChar()
+        Byte::class.java -> 0.toByte()
+        Short::class.java -> 0.toShort()
+        Int::class.java -> 0
+        Float::class.java -> 0f
+        Long::class.java -> 0L
+        Double::class.java -> 0.0
+        Void.TYPE -> throw IllegalStateException("Parameter with void type is illegal")
+        else -> throw UnsupportedOperationException("Unknown primitive: $type")
+      }
+    } else null
+
+private fun Class<*>.toKmClass(throwOnNotClass: Boolean): KmClass? {
+  val metadata = getAnnotation(KOTLIN_METADATA) ?: return null
+  val header = with(metadata) {
+    KotlinClassHeader(
+        kind = kind,
+        metadataVersion = metadataVersion,
+        bytecodeVersion = bytecodeVersion,
+        data1 = data1,
+        data2 = data2,
+        extraString = extraString,
+        packageName = packageName,
+        extraInt = extraInt
+    )
+  }
+  val classMetadata = KotlinClassMetadata.read(header)
+  if (classMetadata !is KotlinClassMetadata.Class) {
+    if (throwOnNotClass) {
+      throw IllegalStateException("Cannot serialize class with metadata $classMetadata")
+    } else {
+      return null
+    }
+  }
+
+  return classMetadata.toKmClass()
+}
+
+private fun Class<*>.allMethods(): Sequence<Method> {
+  return declaredMethods.asSequence() + methods.asSequence()
+}
+
+internal data class ParameterData(
+    val km: KmValueParameter,
+    val index: Int,
+    val rawType: Class<*>,
+    val annotations: List<Annotation>
+) {
+  val name get() = km.name
+  val isOptional get() = Flag.ValueParameter.DECLARES_DEFAULT_VALUE(km.flags)
```

Comment:
Note that semantics of "is optional" are bit different from "declares default value". Not sure how important it is for your use case:
```
interface A {
    fun foo(x: String = "...")
}

interface B : A {
    override fun foo(x: String)
}
```
`x` in `B.foo` is optional, although it does _not_ declare default value.

[`KParameter.isOptional`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.reflect/-k-parameter/is-optional.html) implements the "is optional" semantics.

Reply:
Interesting, that's a good point. I think we can stick with just our default check since this is only ever checked on constructor parameters. Updated the naming of this though to reflect the default value semantic though! 164bd08